### PR TITLE
Update docs to reflect that the largest supported service subnet

### LINKED ISSUE #2132: document the usage of --service-cidr and improve validation
BUG REPORT 

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:54:15Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:48:36Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
- **Cloud provider or hardware configuration**:
brightbox
- **OS** (e.g. from /etc/os-release):
Ubuntu 18.04.4 LTS
- **Kernel** (e.g. `uname -a`):
Linux srv-u28wb 4.15.0-99-generic #100-Ubuntu SMP Wed Apr 22 20:32:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

## What happened?

Recurrence of #695 which I can't reopen

There doesn't appear to be a preflight validation of the **service CIDR** net mask so that it is of a size that is acceptable to the API server.

Therefore when you put `serviceSubnet: fd00::/64` into the config you get

```
ubuntu@srv-3316t:~$ sudo more /var/log/containers/kube-apiserver-srv-3316t_kube-system_kube-apiserver-33f4880590bb8d57aa832dfb54e6b8cd6c58b67eab00e6073dbd1fed8e31b135.log 
2020-05-07T13:10:14.562713334Z stderr F Flag --insecure-port has been deprecated, This flag will be removed in a future version.
2020-05-07T13:10:14.563080628Z stderr F I0507 13:10:14.562942       1 server.go:656] external host was not specified, using 2a02:1348:17c:72b9:24:19ff:fef1:cae6
2020-05-07T13:10:14.563355454Z stderr F Error: specified --service-cluster-ip-range is too large
2020-05-07T13:10:14.564709083Z stderr F Usage:
2020-05-07T13:10:14.564723613Z stderr F   kube-apiserver [flags]

```

from the API server

### PR DESCRIPTION
**Pull request by Arvinderpal**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
Update docs to reflect that the largest supported service subnet is /12 for IPv4 and /108 for IPv6. Also add validation of max allowable subnet size to kubeadm which identical to what the apiserver does. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/2132

**Does this PR introduce a user-facing change?**:
```release-note
kube-controller-manager, kube-apiserver and kubeadm now include a more descriptive message on what IPv4 and IPv6 ranges are supported for service subnet. Add related validation in kubeadm and exit on errors early.
```

### CODE REVIEW

**Comment by Arvinderpal** [comment]
@neolit123 Can you please do a first pass on this?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 1]
```diff
@@ -369,8 +369,20 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
...
 }
 
+func validateMaximumServiceSubnetSize(cidr *net.IPNet, maxCIDRBits int) error {
+	// Should be smallish sized cidr, this thing is kept in etcd
+	// bigger cidr (specially those offered by IPv6) will add no value
+	// significantly increase snapshotting time.
+	// NOTE: This is identical to validation performed in the apiserver.
+	var ones, bits = cidr.Mask.Size()
+	if bits-ones > maxCIDRBits {
+		return fmt.Errorf("specified service subnet is too large; for %d-bit addresses, the mask must be >= %d", bits, bits-maxCIDRBits)
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:379]: `errors.Errorf(....`

(errors is from github.com/pkg/errors)

**Comment by neolit123** [comment]
> Docs have been updated to reflect internal limitations on the service subnet size. kubeadm will also enforce this limitation. 

i'd reword to something like:

> kube-controller-manager, kube-apiserver and kubeadm now include a more descriptive message on what IPv4 and IPv6 ranges are supported for service subnet. Add related validation in kubeadm and exit on errors early.

**Comment by neolit123** [comment]
PTAL
/cc @danwinship @aojea 
@kubernetes/sig-api-machinery-pr-reviews 
@kubernetes/sig-network-pr-reviews

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 2]
```diff
@@ -369,8 +369,20 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+func validateMaximumServiceSubnetSize(cidr *net.IPNet, maxCIDRBits int) error {
+	// Should be smallish sized cidr, this thing is kept in etcd
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:373]: i'd reword to a more formal language - e.g.:
> Should be a small CIDR that is kept in etcd...

**Comment by danwinship** [comment]
> Update docs to reflect that the largest supported service subnet is /12 for IPv4 and /108 for IPv6.

So in #90115 (and later #91606) I was trying to get away from just using "larger" and "smaller" when talking about CIDRs because of the ambiguity where larger subnets have smaller prefix lengths and vice-versa. (Especially given the assumption of IPv6 CIDRs normally being /64s, you could easily read "the largest supported service subnet is /108 for IPv6" as meaning "the prefix length must be no larger than /108", which is the opposite of what was meant.)

One way to talk about it is to say that we don't allow more than 20 bits for the service subnet, so with IPv4 the service CIDR must have a prefix length of /12 or greater, and for IPv6 it must have a prefix length of /108 or greater. (But I get that this doesn't really work well in the space available for documenting a command-line option...)

Something to think about at least.

FTR, we don't _really_ allow prefix lengths of /108 - /111 in IPv6; if you specify anything less than /112, it gets silently changed to /112. (So, eg, if you specify `fd02::/108` and then try to manually allocate `clusterIP: fd02::1:0` it will fail, because even though that's in `fd02::/108`, it's not in `fd02::/112`.)

**Comment by Arvinderpal** [comment]
> One way to talk about it is to say that we don't allow more than 20 bits for the service subnet, so with IPv4 the service CIDR must have a prefix length of /12 or greater, and for IPv6 it must have a prefix length of /108 or greater. (But I get that this doesn't really work well in the space available for documenting a command-line option...)
 
I'm okay with wording this in terms of prefix/mask length. Something like:
```Service CIDR must have a prefix length of /12 or greater for IPv4 and /108 or greater for IPv6.```
 
> FTR, we don't _really_ allow prefix lengths of /108 - /111 in IPv6; if you specify anything less than /112, it gets silently changed to /112. (So, eg, if you specify `fd02::/108` and then try to manually allocate `clusterIP: fd02::1:0` it will fail, because even though that's in `fd02::/108`, it's not in `fd02::/112`.)

Yes, I have noticed this behavior as well, but wanted to keep the kubeadm validation consistent with apiserver, so went with /108 instead of /112. 

Unless some significant changes are made in the underlying code, I don't think we'll have a perfect solution here. However, having clear docs would definitely help the users.

**Comment by neolit123** [comment]
@danwinship there is desire for a short term update to validation / docs to reflect what we actually support ATM. potential amends to the supported CIDR ranges can happen in subsequent PRs.

**Comment by danwinship** [comment]
Right, I was linking to that open PR just to point out the parts where I was fiddling with the wording of error messages.

The other point I made about /108 vs /112 in IPv6 applies to the current code. In 1.18, if you say `--service-cidr-ip-range fd02::/108` you will actually get `fd02::/112`.

**Comment by neolit123** [comment]
if we are updating documentation maybe we should be clear that the numbers are not accurate in the current version?

**Comment by Arvinderpal** [comment]
Would it be sufficient if I made it:
```Currently, the largest supported service subnet is /12 for IPv4 and /112 for IPv6.```
Instead of the `/108`?
Don't think we should make the docs too verbose.

**Comment by sftim** [comment]
Should this documentation change also be made on the website?

File: cmd/kube-apiserver/app/options/options.go [hunk 3]
```diff
@@ -192,7 +192,7 @@ func (s *ServerRunOptions) Flags() (fss cliflag.NamedFlagSets) {
 	// TODO (khenidak) change documentation as we move IPv6DualStack feature from ALPHA to BETA
 	fs.StringVar(&s.ServiceClusterIPRanges, "service-cluster-ip-range", s.ServiceClusterIPRanges, ""+
 		"A CIDR notation IP range from which to assign service cluster IPs. This must not "+
-		"overlap with any IP ranges assigned to nodes or pods.")
+		"overlap with any IP ranges assigned to nodes or pods. Currently, the largest supported service subnet is /12 for IPv4 and /108 for IPv6.")
```
Comment (sftim) [inline cmd/kube-apiserver/app/options/options.go:195]:
```suggestion
		"overlap with any IP ranges assigned to nodes or pods. The largest supported service subnet is /12 for IPv4 and /108 for IPv6.")
```
?

“Currently” doesn't really add anything there. It's good to track improvements but docs aren't the best place to track them.

**Comment by Queuecumber** [comment]
Is there anyplace I can read about why we have this restriction? I was planning to use a /64 for my IPv6 service CIDR, since /64 is the standard prefix length in ipv6, and was surprised to see it fail.

File: cmd/kube-apiserver/app/options/options.go (hunk 3 above)
Comment (Arvinderpal) [reply cmd/kube-apiserver/app/options/options.go:195]: @sftim Thank you for the review. If you don't mind, I would like to keep "currently" as I see it being used in other places as well. Also, it suggests to the user that this may not be the case going forward as we do have open issues seeking to address this.

**Comment by Arvinderpal** [comment]
> Should this documentation change also be made on the website?

Definitely, are you referring to for example: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ 
I assume those docs are updated auto-magically by some script, but please let me know otherwise.

**Comment by Arvinderpal** [comment]
> Would it be sufficient if I made it:
> `Currently, the largest supported service subnet is /12 for IPv4 and /112 for IPv6.`
> Instead of the `/108`?
> Don't think we should make the docs too verbose.

@danwinship @neolit123 Any opinions on this?

**Comment by sftim** [comment]
The main place to document IPv6 limits is probably https://kubernetes.io/docs/concepts/services-networking/dual-stack/
Strictly speaking you can set up single-stack IPv6 but I suspect that's not as common, and there isn't really anywhere else that's a good home.

You _could_ add new sections https://kubernetes.io/docs/reference/networking/ipv4/ and https://kubernetes.io/docs/reference/networking/ipv6/ and fill them out - but that's a lot of work to document one snag!

**Comment by neolit123** [comment]
> Any opinions on this?

outlining the limitations, SGTM.

**Comment by aojea** [comment]
I'm working in removing this limitation for the service subnet, do you mind holding this some time, I think it may be possible

**Comment by neolit123** [comment]
> I'm working in removing this limitation for the service subnet, do you mind holding this some time, I think it may be possible

feature freeze is in 9 days, so any major changes might net get enough review until then.

**Comment by aojea** [comment]
> > I'm working in removing this limitation for the service subnet, do you mind holding this some time, I think it may be possible
> 
> feature freeze is in 9 days, so any major changes might net get enough review until then.

it's not a big change IMHO, anyway, I'll send the KEP today and you can decide after that

**Comment by Queuecumber** [comment]
@aojea would you mind CCing me on the kep? I'm interested in reading what you came up with

**Comment by neolit123** [comment]
> it's not a big change IMHO, anyway, I'll send the KEP today and you can decide after that

well the KEP deadline passed and KEPs require an exception in front of SIG Release. i'd say it's unlikely to get this exception in for 1.19, but deferring to SIG Network on this topic.

what i wanted is to document what is supported ASAP, because this topic is lacking clarity and the users are confused.
if what is supported changes, we can change the documentation too.

**Comment by Arvinderpal** [comment]
I would second @neolit123 - we should document the current limitations. This has been lacking for some time now. 
@aojea What is the timeline for your changes?

**Comment by aojea** [comment]
> I would second @neolit123 - we should document the current limitations. This has been lacking for some time now.
> @aojea What is the timeline for your changes?

just go ahead, if something changes we have to update this anyway, ... seems I was too optimistic :)

File: cmd/kubeadm/app/constants/constants.go [hunk 4]
```diff
@@ -195,6 +195,9 @@ const (
 	// We need at least ten, because the DNS service is always at the tenth cluster clusterIP
 	MinimumAddressesInServiceSubnet = 10
 
+	// MaximumBitsForServiceSubnet defines maximum possible size of the service subnet in terms of bits. For example, if value is 20, then the largest supported service subnet is /12 for IPv4 and /108 for IPv6.
+	MaximumBitsForServiceSubnet = 20
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:199]: reading https://github.com/kubernetes/kubernetes/pull/91875#issuecomment-640614529 again.

to my understanding 108 will pass validation, but in reality anything lower than 112 will be clamped to 112 in another component. should we have a comment about that here?

Comment (Arvinderpal) [reply cmd/kubeadm/app/constants/constants.go:199]: Yes, you're correct. I will add a comment here.

**Review by neolit123 — COMMENTED** [review]
/approve

File: cmd/kube-apiserver/app/options/options.go (hunk 3 above)
Comment (sftim) [reply cmd/kube-apiserver/app/options/options.go]: I _think_ we should try to remove “Currently” from those other places, based on the style guide for documentation:
https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future

Comment (Arvinderpal) [reply cmd/kube-apiserver/app/options/options.go]: Done!

**Comment by Arvinderpal** [comment]
Looks like all the tests finally passed.
@neolit123 @aojea @sftim @danwinship  Do you guys want to take a final look at this?

**Comment by sftim** [comment]
No further questions, your honors.

**Comment by neolit123** [comment]
/lgtm from my side; milestone is set.

needs approval for controller-manager and apiserver.

**Comment by nikhita** [comment]
/assign @liggitt 
for approval

@neolit123 just wanted to double check -- this PR has the v1.19 milestone. Is that intentional? Since code freeze will be lifted soon, only blocker bugs are being considered for 1.19.

**Comment by sftim** [comment]
LGTM (also from a SIG Docs point of view)

**Comment by neolit123** [comment]
> just wanted to double check -- this PR has the v1.19 milestone. Is that intentional? Since code freeze will be lifted soon, only blocker bugs are being considered for 1.19.

it did not make it in time for 1.19 and it can't be considered critical.
/milestone v1.20

**Comment by neolit123** [comment]
/milestone clear
see https://groups.google.com/d/msgid/kubernetes-dev/CAOqU-DTXYyBx7Vs%2BtOvGsAOQNFnVLAEzhWm4CotpPZQaToEyyw%40mail.gmail.com?utm_medium=email&utm_source=footer

**Comment by neolit123** [comment]
>  justaugustus modified the milestones: v1.20, v1.20-phase-doc-cleanup 15 hours ago
looks like this already had a milestone.

/milestone v1.20-phase-doc-cleanup

**Comment by mkorbi** [comment]
Hello @Arvinderpal  
This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
As the PR is tagged for 1.20, is it still planned for this release?

**Comment by Arvinderpal** [comment]
@mkorbi Not sure why this never got merged. 
@neolit123 do you know?

**Comment by neolit123** [comment]
@Arvinderpal 

looks like one of the CI jobs got stuck :thinking: 
> pull-kubernetes-e2e-kind-ipv6 Expected — Waiting for status to be reported

/test pull-kubernetes-e2e-kind-ipv6

**Comment by neolit123** [comment]
still needs approval for the core components though.

/approve
for kubeadm

@aojea this might overlap with your PR at https://github.com/kubernetes/kubernetes/pull/95723

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 5]
```diff
@@ -369,8 +369,20 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
...
+	var ones, bits = cidr.Mask.Size()
+	if bits-ones > maxCIDRBits {
+		return errors.Errorf("specified service subnet is too large; for %d-bit addresses, the mask must be >= %d", bits, bits-maxCIDRBits)
+	}
+	return nil
+}
+
 // ValidateIPNetFromString validates network portion of ip address
-func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool, fldPath *field.Path) field.ErrorList {
+func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack, isServiceSubnet bool, fldPath *field.Path) field.ErrorList {
```
Comment (aojea) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:385]: I will not overload current function and add a new one like `ValidateServiceSubnetSize()`
this will allow to unit test it and will be easier to add new code like in 
https://github.com/kubernetes/kubernetes/pull/95723/files#diff-71c615dc50838bd457062893d9f12906d130be2ea2f1ce964317dc43ca82ad2bR407

**Comment by aojea** [comment]
@Arvinderpal I've incorporated the kubeadm changes to this PR https://github.com/kubernetes/kubernetes/pull/95723 keeping you as author, so people in other sigs don't have to approve kubeadm code.
PTAL

**Comment by mkorbi** [comment]
Hey @neolit123 @Arvinderpal  short reminder, code freeze will be by today

**Comment by neolit123** [comment]
hi, @Arvinderpal 

@aojea included some of these changes in his PR for kubeadm's validation so they are no longer needed here. but i'd say we still want the improved messages for the rest of the components. if you can't update this before Thursday it's fine.

**Comment by Arvinderpal** [comment]
@neolit123 I assume you're talking about api and controller-manager doc changes that are in this PR. However, with @aojea changes, does this still apply? 
```The largest supported service subnet is /12 for IPv4 and /112 for IPv6.")```

**Comment by aojea** [comment]
> The largest supported service subnet is /12 for IPv4 and /112 for IPv6.")

yep, those are kubernetes limitations https://github.com/kubernetes/enhancements/pull/1881

**Comment by neolit123** [comment]
> I assume you're talking about api and controller-manager doc changes that are in this PR. However, with @aojea changes, does this still apply?

yes, if someone is not using kubeadm.
so we can update the PR to only include the !kubeadm changes.

also the release note has to be updated, i guess:
> kube-controller-manager, kube-apiserver now include a more descriptive message on what IPv4 and IPv6 ranges are supported for service subnet.

**Comment by sayanchowdhury** [comment]
Hi 👋🏽 I'm from the Bug Triage team. We've crossed the Code Freeze for 1.20 release on 12th November. As this PR is tagged with 1.20, I'm sending a final reminder to either move the milestone to 1.21 or clear the milestone.

**Comment by neolit123** [comment]
moving the milestone:
/milestone v1.21

**Comment by liggitt** [comment]
reassigning to network folks for review, feel free to reassign me once it has lgtm and needs approval

**Comment by bowei** [comment]
Does it make sense to be so specific in the flags usage docs?

I get concerned that at some point this doc will be out of date/out of sync and people will end up being confused.

If we want, it should be fmt.Sprint'd from the limit constants in the code so that the relationship in the code is clear.

**Comment by aojea** [comment]
> Does it make sense to be so specific in the flags usage docs?

this is a recurrent problem that hit users that try to deploy IPv6, and when the cluster fails to install is not clear to them what is the problem, so they open a new issue or ask in slack.
I tried several times to remove the limitation in Kubernetes, that is clearly the right solution, and I hope that next try in 1.22 we can do it.
In the meantime, I don´t have a strong opinion on the flag usage, but I agree with you in

> I get concerned that at some point this doc will be out of date/out of sync and people will end up being confused.

**Comment by erismaster** [comment]
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day, tomorrow the 9th! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

**Comment by erismaster** [comment]
We have past code freeze for 1.21, going to move this to 1.22.

/milestone v1.22

**Comment by frittentheke** [comment]
@Arvinderpal could you kindly rebase your PR for 1.22? It's just sad to not have this properly documented with more and more dual stack support in K8s now.

**Comment by aojea** [comment]
> @Arvinderpal could you kindly rebase your PR for 1.22? It's just sad to not have this properly documented with more and more dual stack support in K8s now.

@frittentheke do you refer to the service subnet limitation for ipv6 or to the dual stack docs?
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/ , the later seems pretty accurate ..

**Comment by frittentheke** [comment]
> > @Arvinderpal could you kindly rebase your PR for 1.22? It's just sad to not have this properly documented with more and more dual stack support in K8s now.
> 
> @frittentheke do you refer to the service subnet limitation for ipv6 or to the dual stack docs?
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/ , the later seems pretty accurate ..

I was referring to the documentation of the largest service subnet of /112 which is not documented explicitly on https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/#create-a-dual-stack-cluster

Background: https://github.com/kubernetes/kubeadm/issues/2132
